### PR TITLE
[FIX] calendar: recurring events have wrong end date

### PR DESCRIPTION
**Pull request by backspac**
[FIX] calendar: recurring events have wrong end date

Steps to reproduce:
- Install Online Appointments
- Go to Calendar > Online Appointments > create a new one
- Add an available employee like "Mitchell Admin"
- Make a slot for each week day at 8:00
- Go to the Calendar of the employee and create a recurrent meeting from 6:00 to 12:00 which repeats everyday for 10 days
- Go to the Online Appointments you created then click the button "Go to Website"

Bug:
You can still make an appointment when the employee is busy.

Explanation:
Recurring events had their stop date always set to one hour after their start date. So, appointments could still be made two hours after the start date.

As seen here: https://github.com/odoo/odoo/blob/ee76b25b003f96073f78420960e216acf14bf821/addons/calendar/models/calendar.py#L47
`with_date` is passed to `timedelta()`. And passing `hours=True` works as if `True` was `1`.

```python
>>> from datetime import timedelta
>>> timedelta(hours=True)
datetime.timedelta(seconds=3600)
```

This fix reproduces the behavior before this commit introduced the bug: https://github.com/odoo/odoo/commit/9920f20e4c7753bc17bea71dea3a90f7de687196
That is, if no duration is given or duration is `0`, one hour is used instead.

This bug probably occured because of the naming of the definition: https://github.com/odoo/odoo/blob/ee76b25b003f96073f78420960e216acf14bf821/addons/calendar/models/calendar.py#L33
In fact, `with_date` can be interpreted as being boolean, and the `False` default isn't helping against that. A possibly better alternative to this could be:

```python
def calendar_id2real_id(calendar_id=None, duration=None):
```

opw:2334449

### CODE REVIEW

**Comment by backspac** [comment]
Ping @nim-odoo

**Comment by backspac** [comment]
@fw-bot up to saas-13.3

**Comment by agr-odoo** [comment]
looks ok but why 2 commits?

**Comment by backspac** [comment]
The first commit is a fix I had to make in order to reproduce the issue. It's not related to the issue in itself.

**Comment by nim-odoo** [comment]
> The first commit is a fix I had to make in order to reproduce the issue. It's not related to the issue in itself.

It still needs a proper description, though :smile:

**Comment by nim-odoo** [comment]
robodoo rebase-ff

**Comment by nim-odoo** [comment]
Might worth pointing this:
https://github.com/odoo/odoo/blob/ee76b25b003f96073f78420960e216acf14bf821/addons/calendar/models/calendar.py#L47

which actually works with `True`:
```py
>>> from datetime import timedelta
>>> from datetime import datetime as dt

>>> dt.now() + timedelta(hours=True)
datetime.datetime(2020, 11, 18, 15, 19, 12, 777536)
```

This is actually the source of your problem.

File: addons/calendar/models/calendar.py [hunk 1]
```diff
@@ -1769,7 +1769,7 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
         new_args = []
         for arg in args:
             new_arg = arg
-            if arg[0] in ('stop_date', 'stop_datetime', 'stop',) and arg[1] == ">=":
+            if arg[0] in ('stop_date', 'stop_datetime', 'stop',) and arg[1] in ('>=', '>', '=',):
```
Comment (nim-odoo) [inline addons/calendar/models/calendar.py:1772]: still unsure this is necessary

**Review by nim-odoo — APPROVED** [review]
robodoo delegate+ rebase-ff

**Comment by backspac** [comment]
robodoo r+

**Pull request closed without merging**